File: src/lib/embeddings.ts
/**
 * Embedding generation using HuggingFace Inference API with BGE-M3 model.
 * Supports multilingual text (English, French, Arabic).
 */

const EMBEDDING_API_URL =
  process.env.EMBEDDING_API_URL ||
  "https://api-inference.huggingface.co/models/BAAI/bge-m3";

export async function generateEmbedding(text: string): Promise<number[]> {
  const token = process.env.HF_API_TOKEN;
  if (!token) {
    throw new Error("Missing HF_API_TOKEN environment variable");
  }

  const response = await fetch(EMBEDDING_API_URL, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      inputs: text,
      options: { wait_for_model: true },
    }),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(
      `Embedding API error (${response.status}): ${errorText}`
    );
  }

  const result = await response.json();

  // HuggingFace returns either a flat array or nested array
  if (Array.isArray(result) && Array.isArray(result[0])) {
    return result[0] as number[];
  }
  return result as number[];
}

export async function generateEmbeddings(
  texts: string[]
): Promise<number[][]> {
  const token = process.env.HF_API_TOKEN;
  if (!token) {
    throw new Error("Missing HF_API_TOKEN environment variable");
  }

  const response = await fetch(EMBEDDING_API_URL, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      inputs: texts,
      options: { wait_for_model: true },
    }),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(
      `Embedding API error (${response.status}): ${errorText}`
    );
  }

  const result = await response.json();
  return result as number[][];
}

File: backend/app/services/embeddings.py
"""Embedding generation using HuggingFace Inference API (BGE-M3).

Uses the Inference Providers router endpoint:
https://router.huggingface.co/hf-inference/models/<model>/pipeline/feature-extraction
"""

import httpx
from app.config import settings


def _to_sentence_vector(result) -> list[float]:
    """Normalize a feature-extraction response into a single 1D embedding.

    The pipeline can return:
      - a 1D array (pooled sentence embedding) -> use as-is
      - a 2D array (per-token embeddings)      -> mean-pool over tokens
    """
    if not isinstance(result, list) or len(result) == 0:
        raise RuntimeError(f"Unexpected embedding response: {repr(result)[:200]}")

    # 1D: list of floats
    if isinstance(result[0], (int, float)):
        return [float(x) for x in result]

    # 2D: list of token vectors -> mean pool
    if isinstance(result[0], list):
        num_tokens = len(result)
        dim = len(result[0])
        pooled = [0.0] * dim
        for token_vec in result:
            for i, v in enumerate(token_vec):
                pooled[i] += float(v)
        return [v / num_tokens for v in pooled]

    raise RuntimeError(f"Unexpected embedding response shape: {repr(result)[:200]}")


async def generate_embedding(text: str) -> list[float]:
    """Generate embedding for a single text."""
    if not settings.hf_api_token:
        raise ValueError("Missing HF_API_TOKEN environment variable")

    async with httpx.AsyncClient(timeout=60.0) as client:
        response = await client.post(
            settings.embedding_api_url,
            headers={"Authorization": f"Bearer {settings.hf_api_token}"},
            json={"inputs": text},
        )

    if response.status_code != 200:
        raise RuntimeError(
            f"Embedding API error ({response.status_code}): {response.text}"
        )

    return _to_sentence_vector(response.json())


async def generate_embeddings(texts: list[str]) -> list[list[float]]:
    """Generate embeddings for multiple texts."""
    if not settings.hf_api_token:
        raise ValueError("Missing HF_API_TOKEN environment variable")

    async with httpx.AsyncClient(timeout=120.0) as client:
        response = await client.post(
            settings.embedding_api_url,
            headers={"Authorization": f"Bearer {settings.hf_api_token}"},
            json={"inputs": texts},
        )

    if response.status_code != 200:
        raise RuntimeError(
            f"Embedding API error ({response.status_code}): {response.text}"
        )

    result = response.json()

    # Batch response is a list, one entry per input text. Each entry may be
    # a 1D pooled vector or a 2D token matrix.
    return [_to_sentence_vector(item) for item in result]
